### PR TITLE
Fix race condition issue in data contract serializers

### PR DESCRIPTION
**Pull request by khdang**
Use ConcurrentDictionary instead of Dictionary. 
I ran perf tests a few times to get the comparison before and after the change. Although there are noises in the results no consistent trend of regression is identified.

Fix https://github.com/dotnet/corefx/issues/8422
cc: @shmao @SGuyGe @zhenlan

### CODE REVIEW

**Comment by shmao** [comment]
LGTM.

**Comment by justinvp** [comment]
Now that `ConcurrentDictionary` is being used, `GetOrAdd` could be used at the callsite, e.g.:

```csharp
static internal object UnsafeGetUninitializedObject(Type type)
{
    ...

    const BindingFlags flags = BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance;
    bool hasDefaultConstructor = s_typeHasDefaultConstructorMap.GetOrAdd(type, t => t.GetConstructor(flags, Array.Empty<Type>()) != null));

    ....
}
```

**Comment by zhenlan** [comment]
LGTM

**Comment by khdang** [comment]
Updated as @justinvp 's suggestion

File: src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs [hunk 1]
```diff
@@ -871,12 +872,8 @@ static internal object UnsafeGetUninitializedObject(Type type)
                   return Activator.CreateInstance(type);
             }
 
-            bool hasDefaultConstructor;
-            if (!s_typeHasDefaultConstructorMap.TryGetValue(type, out hasDefaultConstructor))
-            {
-                hasDefaultConstructor = type.GetConstructor(BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance, Array.Empty<Type>()) != null;
-                s_typeHasDefaultConstructorMap[type] = hasDefaultConstructor;
-            }
+            const BindingFlags flags = BindingFlags.NonPublic | BindingFlags.Public | BindingFlags.Instance;
```
Comment (stephentoub) [inline src/System.Private.DataContractSerialization/src/System/Runtime/Serialization/XmlFormatReaderGenerator.cs]: Nit: guidelines are to use PascalCasing for consts, even local ones

**Comment by stephentoub** [comment]
As long as we're ok with the new dependency (cc: @weshaggard), LGTM.

**Comment by weshaggard** [comment]
The dependency is OK. However we should call out that we are adding this API back (https://github.com/dotnet/corefx/issues/8133) at which point this extra logic shouldn't be needed at all.

**Comment by stephentoub** [comment]
Sounds like we should proceed with it and then remove it when it's no longer necessary.

**Pull request merged**